# 오류란 무엇인가?

### 컴파일 오류

프로그램 코드 작성 중 발생하는 문법적 오류

### 실행 오류
실행중인 프로그램이 의도하지않은 동작을 하거나(bug) 프로그램이 중지되는 오류(runtime error)

자바는 예외처리를 통해 이러한 프로그램의 비정상 종료를 막고, 로그를 남길수있다.

# 오류와 예외 클래스

### 시스템 오류(error)

가상 머신에서 발생, 프로그래머가 처리 할 수 없음

동적메모리를 다 사용한 경우, stack over flow등

### 예외(Exception)

프로그램에서 제어할 수 있는 오류

읽으려는 파일이 없는경우, 네트워크 연결 오류 등

자바프로그램에서 예외에 대한 처리를 수행함

![image](https://user-images.githubusercontent.com/85108615/198463227-ed3821ca-43fa-4fb1-aa4e-020a4a8437b2.png)


# try-catch문으로 예외 처리하기
```
try{
     예외가 발생 할 수 있는 코드 부분
}catch(처리할 예외 타입 e){
     try블록 안에서 예외가 발생했을 때 수행되는 부분
}
```

# try-catch-finally문으로 예외 처리하기
```
try{
     예외가 발생 할 수 있는 코드 부분
}catch(처리할 예외 타입 e){
     try블록 안에서 예외가 발생했을 때 수행되는 부분
}finally{
     예외 발생 여부와 상관 없이 항상 수행되는 부분
     리소스를 정리하는(끝내는)코드를 주로 씀
}
```


### 예제 

에러를 만들어 테스트해보기.

```java
package com.java.fastcampus.exception;

//테스트를 하기위해 arr에러를 만들어보자.
public class ArrayExceptionTest {

    public static void main(String[] args) {

        int[] arr = new int[5];

        //i가 >=5인데 반대로 써주기기
        for (int i = 0; i <= 5; i++) {
            System.out.println(arr[i]);
        }
    }
}

```

arr에 값을 넣어주지않았지만 0으로 초기화되어있다.

출력결과

![image](https://user-images.githubusercontent.com/85108615/198468310-000ade89-c3f2-4845-bf69-5ae33c0f1f4f.png)


5보다 큰곳을 출력하라고하니, 없다고 에러가 난다 => try-catch문으로 처리해보자.

```java

package com.java.fastcampus.exception;

//테스트를 하기위해 arr에러를 만들어보자.
public class ArrayExceptionTest {

    public static void main(String[] args) {

        int[] arr = new int[5];

        try {
            for (int i = 0; i <= 5; i++) {
                System.out.println(arr[i]);
            }
        } catch (ArrayIndexOutOfBoundsException e) {
                System.out.println(e); //e.toString으로 기본적인 에러에 대한 정보가 출력됨
                System.out.println("예외처리");
        }
        System.out.println("프로그램 종료");
    }
}

```
출력결과

![image](https://user-images.githubusercontent.com/85108615/198469955-a0eb1542-c262-42ed-a491-0148e349cd39.png)

위에서는 에러가 발생해 다운되어버린것이고, 이번에는 다운이아니라 catch에있는 수행문을 전부 수행하고

프로그램이 정상적으로 종료된것.



